File: sections/15_bfs_unweighted_shortest_paths/problems/c_grid_rescue_path/solution.py
from collections import deque
import sys


DIRS = [(1, 0, "D"), (0, -1, "L"), (0, 1, "R"), (-1, 0, "U")]


def main() -> None:
    input = sys.stdin.readline
    n, m = map(int, input().split())
    grid = [list(input().strip()) for _ in range(n)]
    start = goal = (-1, -1)
    for r in range(n):
        for c in range(m):
            if grid[r][c] == "S":
                start = (r, c)
            elif grid[r][c] == "G":
                goal = (r, c)

    dist = [[-1] * m for _ in range(n)]
    parent = [[None] * m for _ in range(n)]
    sr, sc = start
    gr, gc = goal
    dist[sr][sc] = 0
    q = deque([start])
    while q:
        r, c = q.popleft()
        for dr, dc, move in DIRS:
            nr, nc = r + dr, c + dc
            if not (0 <= nr < n and 0 <= nc < m):
                continue
            if grid[nr][nc] == "#" or dist[nr][nc] != -1:
                continue
            dist[nr][nc] = dist[r][c] + 1
            parent[nr][nc] = (r, c, move)
            q.append((nr, nc))

    if dist[gr][gc] == -1:
        print("NO")
        return

    path: list[str] = []
    r, c = goal
    while (r, c) != start:
        pr, pc, move = parent[r][c]
        path.append(move)
        r, c = pr, pc
    path.reverse()
    print("YES")
    print(len(path))
    print("".join(path))


if __name__ == "__main__":
    main()

File: sections/15_bfs_unweighted_shortest_paths/problems/c_grid_rescue_path/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Parent {
    int r = -1;
    int c = -1;
    char move = '?';
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;
    vector<string> grid(n);
    pair<int, int> start{-1, -1}, goal{-1, -1};
    for (int r = 0; r < n; r++) {
        cin >> grid[r];
        for (int c = 0; c < m; c++) {
            if (grid[r][c] == 'S') start = {r, c};
            if (grid[r][c] == 'G') goal = {r, c};
        }
    }

    const vector<int> dr = {1, 0, 0, -1};
    const vector<int> dc = {0, -1, 1, 0};
    const vector<char> mv = {'D', 'L', 'R', 'U'};

    vector<vector<int>> dist(n, vector<int>(m, -1));
    vector<vector<Parent>> parent(n, vector<Parent>(m));
    queue<pair<int, int>> q;
    dist[start.first][start.second] = 0;
    q.push(start);
    while (!q.empty()) {
        auto [r, c] = q.front();
        q.pop();
        for (int i = 0; i < 4; i++) {
            int nr = r + dr[i];
            int nc = c + dc[i];
            if (nr < 0 || nr >= n || nc < 0 || nc >= m) continue;
            if (grid[nr][nc] == '#' || dist[nr][nc] != -1) continue;
            dist[nr][nc] = dist[r][c] + 1;
            parent[nr][nc] = {r, c, mv[i]};
            q.push({nr, nc});
        }
    }

    auto [gr, gc] = goal;
    if (dist[gr][gc] == -1) {
        cout << "NO\n";
        return 0;
    }

    string path;
    for (pair<int, int> cur = goal; cur != start;) {
        Parent p = parent[cur.first][cur.second];
        path.push_back(p.move);
        cur = {p.r, p.c};
    }
    reverse(path.begin(), path.end());
    cout << "YES\n" << path.size() << '\n' << path << '\n';
    return 0;
}

